### PR TITLE
Assignment8

### PR DESCRIPTION
**Pull request by NicholasJohnsonDev**
Assignment 8

### CODE REVIEW

**Comment by NicholasJohnsonDev** [comment]
Currently working on this, should be done by 2pm

**Comment by Fervidusletum** [comment]
Reviewing

**Review by Fervidusletum — CHANGES_REQUESTED** [review]
## Assignment
- Update the `GroupRepository` with functionality to generate gift assignments. ✔
  - A group with with 2 or fewer users should result in an error. This error should be displayed to a user. ✔
  - A user is not allowed to be both the Giver and Recipient of the assignment. ✔
- Update the GroupsController with an endpoint for generating the assignments. ✔
- Update the Groups/Edit page to contain the UI elements for the following
  - A means to generate assignments. ✔
  - A means to display any errors that occurs while generating assignments. ✔
  - Display of the assignments for each of the users. ✔
- Write appropriate unit tests for changes made to the `GroupRepository` ✔
- Write appropriate functional tests for the change made to the `GroupsController` ❌

## Extra Credit
- There is currently no way to for a user to add gifts. On the edit user page add functionality to be able to add a gift to the user. ❌
- Allow for navigating to a user from the Group edit page to make it easy to see the gifts requested by a user. ❌

## Comments
- The GroupRepository GenerateAssignments_WithInvalidId test intermittently fails.
  - I ran into a similar issue; I fixed it by changing the test value from 42 to a unique value.
- Both GroupController tests for CreateAssignments are failing.
  - WithError expects a badrequest, but gets a notfound.
  - WithSuccess expects an OK but gets a notfound.
- GenerateAssignments isn't tested with regard to ordering of generated assignments.
- While the alert works for displaying errors, it's rather annoying. IMO it would be better to display the error in-line on the page.

File: SecretSanta/src/SecretSanta.Business/GroupRepository.cs [hunk 1]
```diff
@@ -36,6 +37,51 @@ public bool Remove(int id)
...
+        public AssignmentResult GenerateAssignments(int id)
+        {
+
+            if (!MockData.Groups.ContainsKey(id))
+            {
+                return AssignmentResult.Error("Group not found");
+            }
+            else if (MockData.Groups[id].Users.Count() <= 2)
+            {
+                return AssignmentResult.Error("Group Group must have at least three users");
```
Comment (Keboo) [inline SecretSanta/src/SecretSanta.Business/GroupRepository.cs:64]:
```suggestion
                return AssignmentResult.Error("Group must have at least three users");
```

File: SecretSanta/test/SecretSanta.Business.Tests/GroupRepositoryTests.cs [hunk 2]
```diff
@@ -122,6 +122,26 @@ public void GenerateAssignments_WithInvalidId_ReturnsError()
             Assert.AreEqual("Group not found", result.ErrorMessage);
         }
 
+        [TestMethod]
+        public void GenerateAssignments_WithInvalidId_CreatesAssignments()
```
Comment (Keboo) [inline SecretSanta/test/SecretSanta.Business.Tests/GroupRepositoryTests.cs:126]:
```suggestion
        public void GenerateAssignments_WithInvalidId_ReturnsNotFoundError()
```

**Review by Keboo — COMMENTED** [review]
# Assignment
- Update the `GroupRepository` with functionality to generate gift assignments. ✔
  - A group with with 2 or fewer users should result in an error. This error should be displayed to a user. ✔
  - A user is not allowed to be both the Giver and Recipient of the assignment. ✔
- Update the GroupsController with an endpoint for generating the assignments. ✔
- Update the Groups/Edit page to contain the UI elements for the following
  - A means to generate assignments. ✔
  - A means to display any errors that occurs while generating assignments. ✔
  - Display of the assignments for each of the users. ✔
- Write appropriate unit tests for changes made to the `GroupRepository` ✔
- Write appropriate functional tests for the change made to the `GroupsController` ✔

**Comment by Keboo** [comment]
# Assignment
- Update the `GroupRepository` with functionality to generate gift assignments. ✔
  - A group with with 2 or fewer users should result in an error. This error should be displayed to a user. ✔
  - A user is not allowed to be both the Giver and Recipient of the assignment. ✔
- Update the GroupsController with an endpoint for generating the assignments. ✔
- Update the Groups/Edit page to contain the UI elements for the following
  - A means to generate assignments. ✔
  - A means to display any errors that occurs while generating assignments. ✔
  - Display of the assignments for each of the users. ✔
- Write appropriate unit tests for changes made to the `GroupRepository` ✔
- Write appropriate functional tests for the change made to the `GroupsController` ✔

**Pull request closed without merging**